File: src/hubs/hub4_veterinary_operations/avian_exotic_dosage_guard.py
"""Avian/Exotic Dosage Guard (Agent 13, Hub 4)."""

import hashlib
import os
from datetime import datetime, timezone

SUBJECT_ID_PEPPER = os.environ.get("SUBJECT_ID_PEPPER", "CHANGE_ME_IN_PRODUCTION")

RX_ALLOWLIST = ["species", "weight_grams", "drug_name", "requested_dosage_mg"]

# Mock micro-dosing safety ceilings for sub-1000g (avian/reptilian) patients.
MICRO_DOSING_SAFETY_THRESHOLDS = {
    "meloxicam": {"max_micrograms_per_gram": 0.2},
    "enrofloxacin": {"max_micrograms_per_gram": 10},
    "ivermectin": {"max_micrograms_per_gram": 0.2},
}


def hash_subject_id(owner_name, pet_name):
    return hashlib.sha256(f"{SUBJECT_ID_PEPPER}:{owner_name}:{pet_name}".encode()).hexdigest()


def mask_value(value):
    s = str(value) if value is not None else ""
    if len(s) <= 2:
        return "*" * len(s)
    return f"{s[0]}{'*' * (len(s) - 2)}{s[-1]}"


def iso_now():
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


class AvianExoticDosageGuard:
    # Stage 1 (Prescription Ingestion): owner/pet identifiers are only used
    # to derive a hashed subject_id; species, weight, drug, and dose survive.
    def ingest(self, raw_request, patient_identifiers):
        patient_identifiers = patient_identifiers or {}
        owner_name = patient_identifiers.get("owner_name")
        pet_name = patient_identifiers.get("pet_name")
        if not owner_name or not pet_name:
            raise ValueError("patient_identifiers requires owner_name and pet_name to derive a subject_id")
        raw_request = raw_request or {}
        species = raw_request.get("species")
        weight_grams = raw_request.get("weight_grams")
        drug_name = raw_request.get("drug_name")
        requested_dosage_mg = raw_request.get("requested_dosage_mg")
        if not species or not drug_name or weight_grams is None or requested_dosage_mg is None:
            raise ValueError("raw_request requires species, weight_grams, drug_name, and requested_dosage_mg")

        record = {
            "subject_id": hash_subject_id(owner_name, pet_name),
            "species": species,
            "weight_grams": float(weight_grams),
            "drug_name": drug_name,
            "requested_dosage_mg": float(requested_dosage_mg),
        }

        for key in list(record.keys()):
            if key != "subject_id" and key not in RX_ALLOWLIST:
                del record[key]
        return record

    # Stage 2 (Metabolic/Species Validation): defense-in-depth guard, then
    # apply the micro-dosing safety ceiling for sub-1000g patients. Fails
    # closed if the patient is a micro-patient but the drug has no threshold
    # on file, routing it to manual pharmacist review rather than approving.
    def validate_dosage(self, record):
        leaked = next((k for k in record if k != "subject_id" and k not in RX_ALLOWLIST), None)
        if leaked:
            raise ValueError(f'Compliance violation: unexpected field "{leaked}" reached validation stage')

        is_micro_patient = record["weight_grams"] < 1000
        threshold_entry = MICRO_DOSING_SAFETY_THRESHOLDS.get(record["drug_name"].lower())
        breach = False
        reason = None
        requested_micrograms_per_gram = None

        if is_micro_patient and threshold_entry:
            requested_micrograms_per_gram = (record["requested_dosage_mg"] * 1000) / record["weight_grams"]
            if requested_micrograms_per_gram > threshold_entry["max_micrograms_per_gram"]:
                breach = True
                reason = (
                    f'Requested dose ({requested_micrograms_per_gram:.2f} mcg/g) exceeds the micro-dosing safety ceiling of '
                    f'{threshold_entry["max_micrograms_per_gram"]} mcg/g for {record["drug_name"]} in a {record["weight_grams"]}g patient.'
                )
        elif is_micro_patient and not threshold_entry:
            breach = True
            reason = f'No micro-dosing safety threshold on file for "{record["drug_name"]}" in a sub-1000g patient — route to manual pharmacist review.'

        return {
            "is_micro_patient": is_micro_patient,
            "requested_micrograms_per_gram": requested_micrograms_per_gram,
            "breach": breach,
            "reason": reason,
            "cleared": not breach,
        }

    # Stage 3 (Integration): only reachable when Stage 2 clears the dose.
    def build_compounding_label(self, record):
        now = datetime.now(timezone.utc)
        return (
            f'CLEANROOM-LABEL|{record["subject_id"]}|SPECIES:{record["species"]}|DRUG:{record["drug_name"]}|'
            f'DOSE_MG:{record["requested_dosage_mg"]}|WEIGHT_G:{record["weight_grams"]}|VERIFIED:TRUE|'
            f'LABEL_ID:CL-{record["subject_id"][:10]}-{int(now.timestamp() * 1000)}'
        )

    # Stage 4a (Micro-Dosing Intercept): only reachable on a validation
    # breach — the compounding label is never generated on this path.
    def build_micro_dosing_intercept(self, record, validation_result):
        return {
            "channel": "pharmacist_micro_dosing_intercept",
            "priority": "active_warning",
            "subject_id": record["subject_id"],
            "drug_name": record["drug_name"],
            "headline": "Micro-dosing intercept: catastrophic overdose risk in sub-1000g patient",
            "reason": validation_result["reason"],
            "requested_action": "Do not compound or dispense as ordered. Recalculate dose against a verified exotic-species formulary before proceeding.",
            "created_at": iso_now(),
        }

    # Stage 4b (Audit Telemetry): tokenized Splunk log — the drug name is
    # masked before it ever reaches the log.
    def build_splunk_event(self, record, validation_result, outcome):
        return {
            "time": datetime.now(timezone.utc).timestamp(),
            "host": "avian-exotic-dosage-guard",
            "source": "hub4_veterinary_operations/avian_exotic_dosage_guard",
            "sourcetype": "_json",
            "event": {
                "subject_id": record["subject_id"],
                "action": outcome,
                "drug_name": mask_value(record["drug_name"]),
                "is_micro_patient": validation_result["is_micro_patient"],
                "breach": validation_result["breach"],
                "requested_micrograms_per_gram": validation_result["requested_micrograms_per_gram"],
                "processed_at": iso_now(),
            },
        }

    def run(self, raw_request, patient_identifiers):
        record = self.ingest(raw_request, patient_identifiers)
        validation_result = self.validate_dosage(record)

        if validation_result["cleared"]:
            compounding_label = self.build_compounding_label(record)
            splunk_event = self.build_splunk_event(record, validation_result, "cleared_for_compounding")
            return {"compounding_label": compounding_label, "intercept": None, "splunk_event": splunk_event}

        intercept = self.build_micro_dosing_intercept(record, validation_result)
        splunk_event = self.build_splunk_event(record, validation_result, "blocked_micro_dosing_exception")
        return {"compounding_label": None, "intercept": intercept, "splunk_event": splunk_event}

File: src/hubs/hub4_veterinary_operations/avian_exotic_dosage_guard.js
'use strict';

const crypto = require('crypto');

const SUBJECT_ID_PEPPER = process.env.SUBJECT_ID_PEPPER || 'CHANGE_ME_IN_PRODUCTION';

const RX_ALLOWLIST = ['species', 'weightGrams', 'drugName', 'requestedDosageMg'];

// Mock micro-dosing safety ceilings for sub-1000g (avian/reptilian) patients.
const MICRO_DOSING_SAFETY_THRESHOLDS = {
  meloxicam: { maxMicrogramsPerGram: 0.2 },
  enrofloxacin: { maxMicrogramsPerGram: 10 },
  ivermectin: { maxMicrogramsPerGram: 0.2 },
};

function hashSubjectId(ownerName, petName) {
  return crypto
    .createHash('sha256')
    .update(`${SUBJECT_ID_PEPPER}:${ownerName}:${petName}`)
    .digest('hex');
}

function maskValue(value) {
  const str = String(value ?? '');
  if (str.length <= 2) return '*'.repeat(str.length);
  return `${str[0]}${'*'.repeat(str.length - 2)}${str[str.length - 1]}`;
}

class AvianExoticDosageGuard {
  // Stage 1 (Prescription Ingestion): owner/pet identifiers are only used
  // to derive a hashed subjectId; species, weight, drug, and dose survive.
  ingest(rawRequest, patientIdentifiers) {
    const { ownerName, petName } = patientIdentifiers || {};
    if (!ownerName || !petName) {
      throw new Error('patientIdentifiers requires ownerName and petName to derive a subjectId');
    }
    const { species, weightGrams, drugName, requestedDosageMg } = rawRequest || {};
    if (!species || !drugName || weightGrams == null || requestedDosageMg == null) {
      throw new Error('rawRequest requires species, weightGrams, drugName, and requestedDosageMg');
    }

    const record = {
      subjectId: hashSubjectId(ownerName, petName),
      species,
      weightGrams: Number(weightGrams),
      drugName,
      requestedDosageMg: Number(requestedDosageMg),
    };

    for (const key of Object.keys(record)) {
      if (key !== 'subjectId' && !RX_ALLOWLIST.includes(key)) {
        delete record[key];
      }
    }
    return record;
  }

  // Stage 2 (Metabolic/Species Validation): defense-in-depth guard, then
  // apply the micro-dosing safety ceiling for sub-1000g patients. Fails
  // closed if the patient is a micro-patient but the drug has no threshold
  // on file, routing it to manual pharmacist review rather than approving.
  validateDosage(record) {
    const leaked = Object.keys(record).find((key) => key !== 'subjectId' && !RX_ALLOWLIST.includes(key));
    if (leaked) {
      throw new Error(`Compliance violation: unexpected field "${leaked}" reached validation stage`);
    }

    const isMicroPatient = record.weightGrams < 1000;
    const thresholdEntry = MICRO_DOSING_SAFETY_THRESHOLDS[record.drugName.toLowerCase()];
    let breach = false;
    let reason = null;
    let requestedMicrogramsPerGram = null;

    if (isMicroPatient && thresholdEntry) {
      requestedMicrogramsPerGram = (record.requestedDosageMg * 1000) / record.weightGrams;
      if (requestedMicrogramsPerGram > thresholdEntry.maxMicrogramsPerGram) {
        breach = true;
        reason = `Requested dose (${requestedMicrogramsPerGram.toFixed(2)} mcg/g) exceeds the micro-dosing safety ceiling of ${thresholdEntry.maxMicrogramsPerGram} mcg/g for ${record.drugName} in a ${record.weightGrams}g patient.`;
      }
    } else if (isMicroPatient && !thresholdEntry) {
      breach = true;
      reason = `No micro-dosing safety threshold on file for "${record.drugName}" in a sub-1000g patient — route to manual pharmacist review.`;
    }

    return { isMicroPatient, requestedMicrogramsPerGram, breach, cleared: !breach };
  }

  // Stage 3 (Integration): only reachable when Stage 2 clears the dose.
  buildCompoundingLabel(record) {
    return `CLEANROOM-LABEL|${record.subjectId}|SPECIES:${record.species}|DRUG:${record.drugName}|DOSE_MG:${record.requestedDosageMg}|WEIGHT_G:${record.weightGrams}|VERIFIED:TRUE|LABEL_ID:CL-${record.subjectId.slice(0, 10)}-${Date.now()}`;
  }

  // Stage 4a (Micro-Dosing Intercept): only reachable on a validation
  // breach — the compounding label is never generated on this path.
  buildMicroDosingIntercept(record, validationResult) {
    return {
      channel: 'pharmacist_micro_dosing_intercept',
      priority: 'active_warning',
      subjectId: record.subjectId,
      drugName: record.drugName,
      headline: 'Micro-dosing intercept: catastrophic overdose risk in sub-1000g patient',
      reason: validationResult.reason,
      requestedAction: 'Do not compound or dispense as ordered. Recalculate dose against a verified exotic-species formulary before proceeding.',
      createdAt: new Date().toISOString(),
    };
  }

  // Stage 4b (Audit Telemetry): tokenized Splunk log — the drug name is
  // masked before it ever reaches the log.
  buildSplunkEvent(record, validationResult, outcome) {
    return {
      time: Date.now() / 1000,
      host: 'avian-exotic-dosage-guard',
      source: 'hub4_veterinary_operations/avian_exotic_dosage_guard',
      sourcetype: '_json',
      event: {
        subjectId: record.subjectId,
        action: outcome,
        drugName: maskValue(record.drugName),
        isMicroPatient: validationResult.isMicroPatient,
        breach: validationResult.breach,
        requestedMicrogramsPerGram: validationResult.requestedMicrogramsPerGram,
        processedAt: new Date().toISOString(),
      },
    };
  }

  run(rawRequest, patientIdentifiers) {
    const record = this.ingest(rawRequest, patientIdentifiers);
    const validationResult = this.validateDosage(record);

    if (validationResult.cleared) {
      const compoundingLabel = this.buildCompoundingLabel(record);
      const splunkEvent = this.buildSplunkEvent(record, validationResult, 'cleared_for_compounding');
      return { compoundingLabel, intercept: null, splunkEvent };
    }

    const intercept = this.buildMicroDosingIntercept(record, validationResult);
    const splunkEvent = this.buildSplunkEvent(record, validationResult, 'blocked_micro_dosing_exception');
    return { compoundingLabel: null, intercept, splunkEvent };
  }
}

module.exports = {
  AvianExoticDosageGuard,
  MICRO_DOSING_SAFETY_THRESHOLDS,
  RX_ALLOWLIST,
};
